File: app/src/main/java/com/itn/securebrowser/util/BlockEngine.java
package com.itn.securebrowser.util;

import java.util.Calendar;
import java.util.List;

public class BlockEngine {

    private final BlockDataStore dataStore;
    private final TimeTracker timeTracker;

    public BlockEngine(BlockDataStore dataStore, TimeTracker timeTracker) {
        this.dataStore = dataStore;
        this.timeTracker = timeTracker;
    }

    public static abstract class BlockReason {
        public static class ScheduleBlock extends BlockReason {}
        public static class LimitReached extends BlockReason {
            public final int limitMinutes;
            public final int usedMinutes;
            public LimitReached(int limitMinutes, int usedMinutes) {
                this.limitMinutes = limitMinutes;
                this.usedMinutes = usedMinutes;
            }
        }
    }

    public BlockReason check(String domain) {
        BlockDataStore.BlockMatch match = dataStore.findDomain(domain);
        if (match == null) return null;

        java.util.HashMap<String, Integer> dailyLimits;
        List<BlockDataStore.BlockSchedule> schedules;

        if (match instanceof BlockDataStore.BlockMatch.GroupMatch) {
            BlockDataStore.BlockGroup g = ((BlockDataStore.BlockMatch.GroupMatch) match).group;
            dailyLimits = g.dailyLimits;
            schedules = g.schedules;
        } else {
            BlockDataStore.BlockSite s = ((BlockDataStore.BlockMatch.SiteMatch) match).site;
            dailyLimits = s.dailyLimits;
            schedules = s.schedules;
        }

        if (isInBlockSchedule(schedules)) return new BlockReason.ScheduleBlock();

        Integer limitMins = dailyLimits.get(todayDayName());
        if (limitMins == null) return null;
        int usedMins = (int) (timeTracker.getTodaySeconds(domain) / 60L);
        if (usedMins >= limitMins) return new BlockReason.LimitReached(limitMins, usedMins);

        return null;
    }

    private boolean isInBlockSchedule(List<BlockDataStore.BlockSchedule> schedules) {
        Calendar cal = Calendar.getInstance();
        String today = dayName(cal.get(Calendar.DAY_OF_WEEK));
        int nowMins = cal.get(Calendar.HOUR_OF_DAY) * 60 + cal.get(Calendar.MINUTE);

        Calendar yCal = Calendar.getInstance();
        yCal.add(Calendar.DAY_OF_YEAR, -1);
        String yesterday = dayName(yCal.get(Calendar.DAY_OF_WEEK));

        for (BlockDataStore.BlockSchedule s : schedules) {
            int from = parseTime(s.from);
            int to = parseTime(s.to);
            if (from <= to) {
                if (s.days.contains(today) && nowMins >= from && nowMins <= to) return true;
            } else {
                if ((s.days.contains(today) && nowMins >= from) ||
                    (s.days.contains(yesterday) && nowMins <= to)) return true;
            }
        }
        return false;
    }

    private int parseTime(String hhmm) {
        String[] parts = hhmm.split(":");
        return Integer.parseInt(parts[0]) * 60 + Integer.parseInt(parts[1]);
    }

    private String todayDayName() {
        return dayName(Calendar.getInstance().get(Calendar.DAY_OF_WEEK));
    }

    private String dayName(int calDay) {
        switch (calDay) {
            case Calendar.SATURDAY: return "SATURDAY";
            case Calendar.SUNDAY: return "SUNDAY";
            case Calendar.MONDAY: return "MONDAY";
            case Calendar.TUESDAY: return "TUESDAY";
            case Calendar.WEDNESDAY: return "WEDNESDAY";
            case Calendar.THURSDAY: return "THURSDAY";
            case Calendar.FRIDAY: return "FRIDAY";
            default: return "MONDAY";
        }
    }

    public static String buildBlockPage(String domain, BlockReason reason) {
        String title, message, icon;
        if (reason instanceof BlockReason.ScheduleBlock) {
            title = "Scheduled block";
            message = "This site is blocked during this time.";
            icon = "\uD83D\uDD50";
        } else {
            BlockReason.LimitReached lr = (BlockReason.LimitReached) reason;
            title = "Daily limit reached";
            message = "You used " + lr.usedMinutes + " of " + lr.limitMinutes + " allowed minutes today.";
            icon = "\u23F1\uFE0F";
        }
        return "<!DOCTYPE html><html dir=\"ltr\" lang=\"en\"><head>" +
            "<meta charset=\"UTF-8\"/><meta name=\"viewport\" content=\"width=device-width,initial-scale=1\"/>" +
            "<title>Blocked</title><style>" +
            "*{margin:0;padding:0;box-sizing:border-box}" +
            "body{min-height:100vh;display:flex;align-items:center;justify-content:center;background:#0f0f0f;font-family:'Segoe UI',Tahoma,sans-serif;color:#e0e0e0}" +
            ".card{background:#1a1a1a;border:1px solid #2a2a2a;border-radius:20px;padding:40px 32px;max-width:360px;width:90%;text-align:center;box-shadow:0 8px 32px rgba(0,0,0,0.5)}" +
            ".icon{font-size:64px;margin-bottom:20px;display:block}" +
            ".title{font-size:22px;font-weight:700;color:#fff;margin-bottom:12px}" +
            ".domain{font-size:14px;color:#888;margin-bottom:20px}" +
            ".message{font-size:15px;color:#aaa;line-height:1.7}" +
            ".divider{width:40px;height:2px;background:linear-gradient(90deg,#444,transparent);margin:24px auto;border-radius:2px}" +
            "</style></head><body><div class=\"card\">" +
            "<span class=\"icon\">" + icon + "</span>" +
            "<div class=\"title\">" + title + "</div>" +
            "<div class=\"domain\">" + domain + "</div>" +
            "<div class=\"divider\"></div>" +
            "<div class=\"message\">" + message + "</div>" +
            "</div></body></html>";
    }
}

File: phase-t1/app/src/main/java/com/itn/securebrowser/BlockEngine.kt
package com.itn.securebrowser

import java.util.Calendar

// ══════════════════════════════════════════════════════════════════════════════
// BlockReason — سبب الحجب (يُمرَّر إلى صفحة الحجب لعرض رسالة مناسبة)
// ══════════════════════════════════════════════════════════════════════════════

sealed class BlockReason {
    /** الوقت الحالي داخل نافذة حظر مجدولة. */
    object ScheduleBlock : BlockReason()

    /** استُهلك الحد اليومي المسموح به. */
    data class LimitReached(
        val limitMinutes: Int,
        val usedMinutes: Int
    ) : BlockReason()
}

// ══════════════════════════════════════════════════════════════════════════════
// BlockEngine — يجيب على سؤال واحد: "هل يجب حجب هذا النطاق الآن؟"
// الترتيب: جدول الحظر أولاً ← ثم الحد اليومي
// ══════════════════════════════════════════════════════════════════════════════

class BlockEngine(
    private val dataStore: BlockDataStore,
    private val timeTracker: TimeTracker
) {

    /**
     * يفحص النطاق ويُرجع سبب الحجب، أو null إذا كان الوصول مسموحاً.
     *
     * استخدام:
     *   val reason = blockEngine.check("youtube.com")
     *   if (reason != null) { // احجب وأظهر صفحة الحجب }
     */
    fun check(domain: String): BlockReason? {
        val match = dataStore.findDomain(domain) ?: return null   // النطاق حر

        val (dailyLimits, schedules) = when (match) {
            is BlockMatch.GroupMatch -> match.group.dailyLimits to match.group.schedules
            is BlockMatch.SiteMatch  -> match.site.dailyLimits  to match.site.schedules
        }

        // ① جدول الحظر — يتجاوز الحد اليومي
        if (isInBlockSchedule(schedules)) return BlockReason.ScheduleBlock

        // ② الحد اليومي
        val limitMins = dailyLimits[todayDayName()] ?: return null   // لا حد اليوم
        val usedMins  = (timeTracker.getTodaySeconds(domain) / 60L).toInt()
        if (usedMins >= limitMins) return BlockReason.LimitReached(limitMins, usedMins)

        return null   // مسموح
    }

    // ── فحص الجداول ───────────────────────────────────────────────────────────

    private fun isInBlockSchedule(schedules: List<BlockSchedule>): Boolean {
        val cal       = Calendar.getInstance()
        val today     = dayName(cal.get(Calendar.DAY_OF_WEEK))
        val nowMins   = cal.get(Calendar.HOUR_OF_DAY) * 60 + cal.get(Calendar.MINUTE)

        // نحتاج أمس لجداول تعبر منتصف الليل (مثل 21:00 → 07:00)
        val yCal      = Calendar.getInstance().apply { add(Calendar.DAY_OF_YEAR, -1) }
        val yesterday = dayName(yCal.get(Calendar.DAY_OF_WEEK))

        return schedules.any { s ->
            val from = parseTime(s.from)
            val to   = parseTime(s.to)

            if (from <= to) {
                // نافذة عادية لا تعبر منتصف الليل: مثل 08:00 → 14:00
                today in s.days && nowMins in from..to
            } else {
                // نافذة تعبر منتصف الليل: مثل 21:00 → 07:00
                // الجزء المسائي: اليوم في الأيام والوقت بعد البداية
                // الجزء الصباحي: أمس في الأيام والوقت قبل النهاية
                (today in s.days && nowMins >= from) ||
                (yesterday in s.days && nowMins <= to)
            }
        }
    }

    // ── مساعدات ───────────────────────────────────────────────────────────────

    /** يحوّل "HH:mm" إلى دقائق منذ منتصف الليل. */
    private fun parseTime(hhmm: String): Int {
        val (h, m) = hhmm.split(":").map { it.toInt() }
        return h * 60 + m
    }

    private fun todayDayName(): String =
        dayName(Calendar.getInstance().get(Calendar.DAY_OF_WEEK))

    private fun dayName(calDay: Int): String = when (calDay) {
        Calendar.SATURDAY  -> "SATURDAY"
        Calendar.SUNDAY    -> "SUNDAY"
        Calendar.MONDAY    -> "MONDAY"
        Calendar.TUESDAY   -> "TUESDAY"
        Calendar.WEDNESDAY -> "WEDNESDAY"
        Calendar.THURSDAY  -> "THURSDAY"
        Calendar.FRIDAY    -> "FRIDAY"
        else               -> "MONDAY"
    }
}
